src/main/java/org/testng/internal/BaseTestMethod.java

```diff
@@ -430,38 +430,59 @@ protected void initGroups(Class<? extends ITestOrConfiguration> annotationClass)
     //
     // Init groups depended upon
     //
-    {
+    initRestOfGroupDependencies(annotationClass);
+  }
+
+  protected void initBeforeAfterGroups(Class<? extends ITestOrConfiguration> annotationClass, String[] groups) {
+    String[] groupsAtMethodLevel = calculateGroupsTouseConsideringValuesAndGroupValues(annotationClass, groups);
+    //@BeforeGroups and @AfterGroups annotation cannot be used at Class level. So its always null
+    String[] groupsAtClassLevel = null;
+    setGroups(getStringArray(groupsAtMethodLevel, groupsAtClassLevel));
+    initRestOfGroupDependencies(annotationClass);
+  }
+
+  private String[] calculateGroupsTouseConsideringValuesAndGroupValues(Class<? extends ITestOrConfiguration> annotationClass, String[] groups) {
+    if (groups == null || groups.length == 0) {
       ITestOrConfiguration annotation = getAnnotationFinder().findAnnotation(getConstructorOrMethod(), annotationClass);
-      ITestOrConfiguration classAnnotation = getAnnotationFinder().findAnnotation(getConstructorOrMethod().getDeclaringClass(), annotationClass);
+      groups = null != annotation ? annotation.getGroups() : null;

```

Comment:
`groups = (null != annotation ? annotation.getGroups() : null);` could help for the readability